client/components/TitleEditor/TitleEditor.tsx

```diff
@@ -0,0 +1,185 @@
+	const { initialValue = '', onChange, isReadOnly = false, ...restProps } = props;
+	const node = useRef<HTMLDivElement>(null);
+	const init = useRef(false);
+	const [focused, setFocused] = useState(false);
+	const sharedProps = {
+		...commonProps,
+		...restProps,
+		className: classNames('title-editor-component', restProps.className),
+		'data-placeholder': restProps.placeholder,
+	};
+
+	const onPaste = useCallback(
+		(event: ClipboardEvent) => {
+			event.preventDefault();
+			const editor = node.current;
+			if (editor === null) return;
+			const selection = getSelection();
+			if (selection === null) return;
+			const range = selection.getRangeAt(0);
+			const html = event.clipboardData.getData('text/html');
+			const dom = parseDom(html);
+			sanitizeDocumentFragment(dom.content);
+			let pasted = false;
+			// execCommand is a deprecated API, but it's still found in most
+			// browsers and is the easist way to support undo/redo without
+			// implementing a fully-controlled editor
+			if ('execCommand' in document) {
+				pasted = document.execCommand('insertHTML', false, dom.innerHTML);
+			}
+			// execCommand is either unsupported or the 'insertHTML' command
+			// failed, so we'll need to insert the sanitized content manually.
+			// Undo won't work here, unfortunately.
+			if (pasted === false) {
+				const target = selection.focusNode;
+				if (!target) return;
+				if (!isChildOf(target, editor)) return;
+				range.deleteContents();
+				range.insertNode(dom.content);
+				range.collapse(false);
+			}
+		},
+		[node],
+	);
+
+	const onKeyDown = useCallback((event) => {
+		if (event.key === 'Enter') {
+			event.preventDefault();
+		} else if (event.key.toLowerCase() === 'u' && event.metaKey) {
+			event.preventDefault();
+		} else if (event.key.toLowerCase() === 'b' && event.metaKey) {
+			event.preventDefault();
+			document.execCommand('Bold', false);
+		} else if (event.key.toLowerCase() === 'i' && event.metaKey) {
+			event.preventDefault();
+			document.execCommand('Italic', false);
+		}
+	}, []);
+
+	// contenteditable inserts a <br> when it contains an element (e.g. <b>,
+	// <em>, etc.) and its contents are cleared.
+	const onInput = useCallback(() => {
+		if (node.current?.textContent === '') {
+			node.current.innerHTML = '';
+		}
+	}, [node]);
+
+	const onFocus = useCallback(() => {
+		setFocused(true);
+	}, []);
+
+	const onBlur = useCallback(() => {
+		onChange?.(node.current?.innerHTML ?? '');
+		setFocused(false);
+	}, [onChange]);
+
+	useEffect(() => {
+		if (init.current) return;
+		if (node.current) {
+			node.current.innerHTML = initialValue;
+			init.current = true;
+			onChange?.(initialValue);
+		}
+	}, [initialValue, onChange]);
+
+	useBeforeUnload(
+		() => !isReadOnly && focused && node.current?.innerHTML !== initialValue,
+		'You have unsaved changes to this Pub. Are you sure you want to navigate away?',
+	);
+
+	if (typeof window === 'undefined' || isReadOnly) {
```

Comment:
FYI we have a `ClientOnly` component that abstracts over this kind of `window` checking logic. Not sure if it's worth using here because this looks pretty clean.